client/src/main/java/org/apache/celeborn/client/read/WorkerPartitionReader.java

```diff
@@ -58,75 +62,113 @@ public class WorkerPartitionReader implements PartitionReader {
       throws IOException {
     fetchMaxReqsInFlight = conf.fetchMaxReqsInFlight();
     results = new LinkedBlockingQueue<>();
+    readableLocations.add(location);
+    if (location.getPeer() != null) {
+      readableLocations.add(location.getPeer());
+    }
     // only add the buffer to results queue if this reader is not closed.
     ChunkReceivedCallback callback =
         new ChunkReceivedCallback() {
           @Override
-          public void onSuccess(int chunkIndex, ManagedBuffer buffer) {
+          public void onSuccess(int chunkIndex, ManagedBuffer buffer, PartitionLocation location) {
             // only add the buffer to results queue if this reader is not closed.
-            synchronized (this) {
-              ByteBuf buf = ((NettyManagedBuffer) buffer).getBuf();
-              if (!closed) {
-                buf.retain();
-                results.add(buf);
-              }
+            ByteBuf buf = ((NettyManagedBuffer) buffer).getBuf();
+            if (!closed.get() && !failedLocations.contains(location)) {
+              buf.retain();
+              results.add(new ChunkData(buf, location));
             }
           }
 
           @Override
-          public void onFailure(int chunkIndex, Throwable e) {
-            String errorMsg = "Fetch chunk " + chunkIndex + " failed.";
-            logger.error(errorMsg, e);
-            exception.set(new IOException(errorMsg, e));
+          public void onFailure(int chunkIndex, PartitionLocation location, Throwable e) {
+            readableLocations.remove(location);
+            if (readableLocations.isEmpty()) {
+              String errorMsg = "Fetch chunk " + chunkIndex + " failed.";
+              logger.error(errorMsg, e);
+              exception.set(new IOException(errorMsg, e));
+            } else {
+              try {
+                synchronized (this) {
+                  if (!failedLocations.contains(location)) {
```

Comment:
so this is the key point of this PR? to avoid try fetch PR on the excluded nodes?

Reply:
The key point of this PR is to fetch whole replica  if one replica is failed.